### PR TITLE
Add a helper function `gh_rate_limit` to extract rate limit information.

### LINKED ISSUE #93: return the rate limit in the responses
Maybe as an attribute. Related: #67

### PR DESCRIPTION
**Pull request by rundel**
Adds the function `gh_rate_limit` which returns basic rate limit information from either a `gh_response` object or a call to `GET /rate_limit`.

Currently returns a list with:
* `limit` -  overall limit (integer)
* `remaining` - remaining requests (integer)
*  `reset` - reset time for limit (datetime)

Fixes #93 and may be helpful for #67.

### CODE REVIEW

File: R/gh_rate_limit.R [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  stopifnot(inherits(response, "gh_response"))
+
+  http_res <- attr(response, "response")
+
+  reset <- as.integer(http_res[["x-ratelimit-reset"]] %||% NA)
+  reset <- as.POSIXct(reset, origin = "1970-01-01")
+
+  list(
+    limit     = as.integer(http_res[["x-ratelimit-limit"]] %||% NA),
+    remaining = as.integer(http_res[["x-ratelimit-remaining"]] %||% NA),
```
Comment (gaborcsardi) [inline R/gh_rate_limit.R]: When you change back the `%||%` definition (see below), you'll probably need to update this to check for zero length vectors, or just use something like 
`c(as.integer(http_res[["x-ratelimit-limit"]], NA))[1]`

File: R/utils.R [hunk 2]
```diff
@@ -11,7 +12,7 @@ compact <- function(x) {
 
 ## from purrr, among other places
 `%||%` <- function(x, y) {
-  if (is.null(x)) {
+  if (is.null(x) || length(x) <= 0) {
```
[the remark this answers is not in the document]
Comment (rundel) [reply R/utils.R]: Ok

File: R/gh_rate_limit.R [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#' Optional argument, if missing a call to "GET /rate_limit" will be made.
+#'
+#' @inheritParams gh
+#'
+#' @return A `list` object containing the overall `limit`, `remaining` limit, and the
+#' limit `reset` time.
+#'
+#' @export
+
+gh_rate_limit = function(response, .token = NULL, .api_url = NULL, .send_headers = NULL) {
```
[the remark this answers is not in the document]
Comment (rundel) [reply R/gh_rate_limit.R]: Sounds good

File: R/gh_rate_limit.R (hunk 1 above)
Comment (rundel) [reply R/gh_rate_limit.R]: Don't think I've seen this pattern before

**Comment by rundel** [comment]
One other wrinkle is that depending on the end point you hit you get slightly different details (since things like search have different rate limits).


```
> r  = gh("/user")
> gh_rate_limit(r)
$limit
[1] 5000
$remaining
[1] 4985
$reset
[1] "2020-01-31 14:59:35 PST"
```

```
> r2 = gh("/search/repositories", q="tidyverse")
> gh_rate_limit(r2)
$limit
[1] 30
$remaining
[1] 29
$reset
[1] "2020-01-31 14:15:26 PST"
```

All of this is available from "/rate_limit" and looks like the possible different options are: `core`, `search`, `graphql`,  `integration_manifest`, `source_import`. Search it the only one that "looks" different. I'm not seeing anything obvious in the response headers about what rate limit you used.

Dealing with search separately via a `gh_search` (#33) may make this less of an issue.

**Comment by gaborcsardi** [comment]
> One other wrinkle is that depending on the end point you hit you get slightly different details (since things like search have different rate limits).

I think we can handle these separately/later, this PR is a good step already. :)

**Comment by gaborcsardi** [comment]
Thanks much!

**Pull request merged**